### PR TITLE
Fix connected component

### PR DESCRIPTION
**Pull request by WeichenXu123**
Fix connected component algorithm implementation:

The algorithm is described in this paper: https://dl.acm.org/doi/pdf/10.1145/2670979.2670997

Summary of the algorithm: essentially, the algorithm is doing "large-star" and "small-star" transformations over the graph,
until convergence (i.e. the transformation doesn't change graph structure),

```
graph = input_graph  # input graph each node must has a unique and comparable label number
while (! converged) {
   graph = large_star(graph)
   graph = small_star(graph)
}
```

**large star** transformation:
For each node U, finding the neighbor nodes V that has larger label than node U, and finding the node M that has the minimal label in the set of `{*node_U_neighbor_node_set, node_U}`, for every node V, generate an edge from V to M, the generated edges compose of the "large star" transformation output graph.

**small star** transformation:
For each node U, finding the neighbor nodes V that has smaller label than node U, and finding the node M that has the minimal label in the set of `node_U_neighbor_node_set`, for every node V, generate an edge from V to M, and additionally, generate an edge of U to M, the generated edges compose of the "small star" transformation output graph.

image

**Convergence critieria:**

Convergence means large star and small star transformation generates identical output graph, and we can prove that, the output graph has the same component connectivity, and the topology of the graph is the overall graph is a union of disjoint stars, one for each connected component, and each star has a center node with the minimal label in its connected component.

To check if the graph reaches convergence, we can check the sum of all nodes' "minimal label value in set {*neighbor_nodes, self_node}", the sum keeps decreasing for each pass of large-star / small-star transformation, once it stops decreasing, it means the graph reaches convergence, and we can get each node's connectivity component ID by calculating its "minimal label value in set {*neighbor_nodes, self_node}".

So, current graphframe connected component algorithm code has one error in convergence checking, I correct it in this PR.

### CODE REVIEW

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 1]
```diff
@@ -356,12 +362,14 @@ object ConnectedComponents extends Logging {
...
-      // test convergence
+      minNbrs1 = minNbrs(ee) // src >= min_nbr
+        .persist(intermediateStorageLevel)
 
+      // test convergence
       // Taking the sum in DecimalType to preserve precision.
       // We use 20 digits for long values and Spark SQL will add 10 digits for the sum.
       // It should be able to handle 200 billion edges without overflow.
-      val (currSum, cnt) = ee.select(sum(col(SRC).cast(DecimalType(20, 0))), count("*")).rdd
+      val (currSum, cnt) = minNbrs1.select(sum(col(MIN_NBR).cast(DecimalType(20, 0))), count("*")).rdd
```
Comment (WeichenXu123) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala:372]: This is the convergence critieria correction.

Ref: Lemma 8 in paper https://dl.acm.org/doi/pdf/10.1145/2670979.2670997

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 2]
```diff
@@ -313,12 +313,18 @@ object ConnectedComponents extends Logging {
 
     var converged = false
     var iteration = 1
-    var prevSum: BigDecimal = null
+
+    // compute min neighbors (including self-min)
+    var minNbrs1: DataFrame = minNbrs(ee) // src >= min_nbr
+      .persist(intermediateStorageLevel)
+
+    var prevSum: BigDecimal = minNbrs1.select(sum(col(MIN_NBR).cast(DecimalType(20, 0))), count("*")).rdd
```
Comment (BenWilson2) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala:321]: Is there any concern with an overflow by capping this at 20?

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 3]
```diff
@@ -313,12 +313,18 @@ object ConnectedComponents extends Logging {
 
     var converged = false
     var iteration = 1
-    var prevSum: BigDecimal = null
+
+    // compute min neighbors (including self-min)
+    var minNbrs1: DataFrame = minNbrs(ee) // src >= min_nbr
```
Comment (BenWilson2) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala:318]: This makes sense to do the first minimization vector connection reduction outside of the while loop. +1

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 4]
```diff
@@ -356,12 +362,14 @@ object ConnectedComponents extends Logging {
 
       ee.persist(intermediateStorageLevel)
 
-      // test convergence
+      minNbrs1 = minNbrs(ee) // src >= min_nbr
+        .persist(intermediateStorageLevel)
```
Comment (BenWilson2) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala:366]: Within the termination of the while loop, should we also explicitly call unpersist() in order to remove the cache references to these pseudo-mutable vars that represent an immutable cache state?

**Comment by BenWilson2** [comment]
Let's get the original reporter to confirm with the customer on their data to validate that the convergence correctness is fixed. Let's also be sure to clean up the persistence (as that is something that they complained about) due to the memory cache holding on to those orphan states that were persisted during iterative convergence. 
Once these are done, LGTM!

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala [hunk 5]
```diff
@@ -313,12 +313,18 @@ object ConnectedComponents extends Logging {
...
-    var prevSum: BigDecimal = null
+
+    // compute min neighbors (including self-min)
+    var minNbrs1: DataFrame = minNbrs(ee) // src >= min_nbr
+      .persist(intermediateStorageLevel)
+
+    var prevSum: BigDecimal = minNbrs1.select(sum(col(MIN_NBR).cast(DecimalType(20, 0))), count("*")).rdd
+      .map { r =>
+        (r.getAs[BigDecimal](0), r.getLong(1))
+      }.first()._1
```
Comment (serena-ruan) [inline src/main/scala/org/graphframes/lib/ConnectedComponents.scala:324]: Can we make this a function and reuse it here and below?

**Review by serena-ruan — COMMENTED** [review]
Do we have test to validate the algorithm is correct?

**Comment by WeichenXu123** [comment]
> Do we have test to validate the algorithm is correct?

I will generate some random graph to test.

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala (hunk 5 above)
Comment (WeichenXu123) [reply src/main/scala/org/graphframes/lib/ConnectedComponents.scala:324]: sure. good idea

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala (hunk 4 above)
Comment (WeichenXu123) [reply src/main/scala/org/graphframes/lib/ConnectedComponents.scala:366]: I will revisit and fix the persist/unpersist operations

File: src/main/scala/org/graphframes/lib/ConnectedComponents.scala (hunk 2 above)
Comment (WeichenXu123) [reply src/main/scala/org/graphframes/lib/ConnectedComponents.scala]: added verification.

**Comment by minerjaime** [comment]
This is great!

**Pull request merged**